packages/webapp/src/components/Task/AddSoilAmendmentProducts/QuantityApplicationRate/useQuantityApplicationRate.ts

```diff
@@ -0,0 +1,185 @@
+import { convert } from '../../../../util/convert-units/convert';
+
+interface UseQuantityApplicationRate {
+  total_area: number;
+  total_area_unit: 'm2' | 'ha' | 'ft2' | 'ac'; // as defined in location_area
+  system: 'metric' | 'imperial';
+  watch: UseFormWatch<TaskProductFormFields>;
+  setValue: UseFormSetValue<TaskProductFormFields>;
+  getValues: UseFormGetValues<TaskProductFormFields>;
+}
+
+export const useQuantityApplicationRate = ({
+  total_area,
+  total_area_unit,
+  system,
+  watch,
+  setValue,
+  getValues,
+}: UseQuantityApplicationRate) => {
+  const application_area = watch(TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED);
+  const application_area_unit = watch(TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED_UNIT);
+  const weight_unit = watch(TASK_PRODUCT_FIELD_NAMES.WEIGHT_UNIT);
+  const volume_unit = watch(TASK_PRODUCT_FIELD_NAMES.VOLUME_UNIT);
+  const application_rate_weight_unit = watch(TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_WEIGHT_UNIT);
+  const application_rate_volume_unit = watch(TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_VOLUME_UNIT);
+
+  const [isWeight, setIsWeight] = useState(() => {
+    const volumeValue = getValues(TASK_PRODUCT_FIELD_NAMES.VOLUME);
+    return isNaN(Number(volumeValue));
+  });
+
+  const toggleMeasure = () => {
+    setIsWeight((prev) => !prev);
+  };
+
+  /* Set initial application area to area total area */
+  useEffect(() => {
+    if (!application_area) {
+      setValue(TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED, total_area);
+    }
+
+    /* update unit of application area and string */
+    onPercentLocationChange();
+  }, []);
+
+  /* Update application area + rate based on percent of location */
+  const onPercentLocationChange = () => {
+    const percent_of_location = getValues(TASK_PRODUCT_FIELD_NAMES.PERCENT_OF_LOCATION_AMENDED);
+
+    const calculatedArea = (total_area * (percent_of_location || 100)) / 100;
+    setValue(TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED, calculatedArea);
+
+    /* set unit of the total area component according to default breakpoints */
+    setValue(
+      TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED_UNIT,
+      /* @ts-ignore */
+      getUnitOptionMap()[getDefaultUnit(location_area, calculatedArea, system).displayUnit],
+      { shouldValidate: true },
+    );
+
+    /* update application rate */
+    onQuantityChange();
+  };
+
+  const onQuantityChange = () => {
+    const weight = getValues(TASK_PRODUCT_FIELD_NAMES.WEIGHT);
+    const weight_unit = getValues(TASK_PRODUCT_FIELD_NAMES.WEIGHT_UNIT);
+    const application_area = getValues(TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED);
+
+    const volume = getValues(TASK_PRODUCT_FIELD_NAMES.VOLUME);
+    const volume_unit = getValues(TASK_PRODUCT_FIELD_NAMES.VOLUME_UNIT);
+
+    if ((weight || weight === 0) && isWeight && weight_unit && application_area) {
+      setValue(
+        TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_WEIGHT,
+        convert(weight / application_area)
```

Comment:
I think its really cool that you extended the convert units ! Its really nice to have an example how to do it as I haven't seen that done before. I am also very curious if you tried the following easy way too and if it was different:
```
 (convert(volume).from('l').to('ml') /  convert(area).from('m2').to('ha')) 
```

Reply:
Thanks for the good points about this previously discussed way not workings --- it would not be easier to handle combined units end edit Unit function. This sets us up well with a nice template for future units with even more constituent parts too.